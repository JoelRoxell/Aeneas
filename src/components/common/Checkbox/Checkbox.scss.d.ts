export const checkbox: string;
export const active: string;
