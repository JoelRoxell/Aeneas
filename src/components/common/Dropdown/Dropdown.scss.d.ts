export const dropdown: string;
export const button: string;
export const icon: string;
export const active: string;
export const list: string;
export const itemWrapper: string;
export const item: string;
export const text: string;
