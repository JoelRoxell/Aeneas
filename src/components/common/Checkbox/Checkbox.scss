.checkbox {
  color: white;
  height: rem(20);
  width: rem(20);
  display: flex;
  align-items: center;
  justify-content: center;
  border: 1px solid #ccc;
  border-radius: rem(4);
  box-shadow: 0 0 rem(10) rgba(0, 0, 0, 0.1);

  input {
    display: none;
  }

  &.active {
    background: $active-color;
  }
}
