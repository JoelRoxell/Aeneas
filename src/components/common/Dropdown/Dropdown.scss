.dropdown {
  display: flex;
  flex-direction: column;
  position: relative;

  .button {
    display: flex;
    align-items: center;
    height: rem(50);
    padding-left: rem(14);
    padding-right: rem(14);

    color: $inactive-color;
    fill: $inactive-color;

    .icon {
      margin-right: rem(10);
    }

    &.active {
      color: $active-color;
      fill: $active-color;
    }

    &:hover {
      color: $active-color;
      fill: $active-color;
    }
  }

  .list {
    height: 0;
    transition: height 1s;
    max-height: rem(200);
    overflow: auto;
    position: absolute;
    top: rem(40);
    background: $background-color;
    width: 100%;
    z-index: 1;
    box-shadow: 0 0 10px rgba(0, 0, 0, 0.2);

    &.active {
      height: fit-content;

      padding: rem(14);
      padding: rem(14);
    }

    .item-wrapper {
      .item {
        display: flex;
        align-items: center;
        margin-bottom: rem(3);

        &:nth-child(1) {
          margin-top: rem(3);
        }

        .text {
          flex: 1;
          margin-left: rem(6);
        }
      }
    }
  }
}
