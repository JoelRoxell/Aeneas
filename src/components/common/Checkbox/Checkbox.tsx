import * as React from 'react'
import * as classNames from 'classnames'

import * as style from './Checkbox.scss'

interface CheckboxProps {
  onChange?: () => any
  selected: boolean
  name?: string
}

function Checkbox(props: CheckboxProps) {
  return (
    <div
      onClick={props.onChange}
      className={classNames(style.checkbox, {
        [style.active]: props.selected
      })}
    >
      <input
        className={style.checkbox}
        type="checkbox"
        checked={props.selected}
        onChange={props.onChange}
        name={props.name}
      />
    </div>
  )
}

export default Checkbox
