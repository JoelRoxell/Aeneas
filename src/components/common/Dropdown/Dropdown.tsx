import * as React from 'react'
import * as classNames from 'classnames'
import * as style from './Dropdown.scss'
import Checkbox from 'components/common/Checkbox'
import Icon from '../Icon'

interface Item {
  value: string | number
  id: string
}
interface DropdownProps {
  title: string
  items: Array<Item>
  onSelect: (item: Item | Item[]) => void
  className?: string
  selected?: string
  multi?: boolean
  icon?: string
}

class Dropdown extends React.Component<DropdownProps> {
  state: { selected: Item | Item[] | null; active: boolean }

  constructor(props: DropdownProps) {
    super(props)

    if (this.props.selected) {
      const item = this.props.items.find(x => x.id === this.props.selected)

      if (item) {
        this.state = {
          selected: item,
          active: false
        }
      }
    } else {
      this.state = { selected: this.props.multi ? [] : null, active: false }
    }
  }

  onClick = (item: Item) => {
    if (this.props.multi) {
      const items = this.state.selected as Item[]

      const index = items.findIndex(x => x.id === item.id)
      const selections = items.concat()

      if (index == -1) {
        selections.push(item)
      } else {
        selections.splice(index, 1)
      }

      this.setState({ selected: selections })
      this.props.onSelect(selections)
    } else {
      this.setState({ selected: item })

      this.props.onSelect(item)
    }
  }

  toggle = (state: boolean) => {
    this.setState({ ...this.state, active: state })
  }

  render() {
    // TODO: should animate and then remove item list, not only hide it.
    return (
      <div
        className={classNames(style.dropdown, this.props.className)}
        onMouseLeave={() => this.toggle(false)}
      >
        <div
          className={classNames(style.button, {
            [style.active]: this.state.active
          })}
          onClick={() => this.toggle(!this.state.active)}
        >
          {this.props.icon && <Icon name={this.props.icon} className={style.icon} />}

          {this.props.title}
        </div>

        <div
          className={classNames(style.list, {
            [style.active]: this.state.active
          })}
        >
          {this.props.items.map((item, i) => {
            let selected = false

            if (Array.isArray(this.state.selected)) {
              selected = this.state.selected.findIndex(x => x.id == item.id) > -1
            } else if (this.state.selected) {
              selected = this.state.selected.id == item.id
            }

            return (
              <div key={i} className={style.itemWrapper}>
                <div className={style.item} onClick={_ => this.onClick(item)}>
                  <Checkbox selected={selected} />

                  <div className={style.text}>{item.value}</div>
                </div>
              </div>
            )
          })}
        </div>
      </div>
    )
  }
}

export default Dropdown
