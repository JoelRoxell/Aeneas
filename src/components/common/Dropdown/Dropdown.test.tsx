import * as React from 'react'
import { mount } from 'enzyme'

import Dropdown from 'components/common/Dropdown'
import { act } from 'react-dom/test-utils'
import Checkbox from '../Checkbox'

describe('Dropdown', () => {
  test('render', () => {
    const wrapper = mount(
      <Dropdown
        items={[{ id: '1', value: 'first' }, { id: '2', value: 'second' }]}
        title="Select something"
        onSelect={item => item}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  test('show / hide', () => {
    const wrapper = mount(
      <Dropdown
        items={[{ id: '1', value: 'first' }, { id: '2', value: 'second' }]}
        title="Select something"
        onSelect={item => console.log('selected', item)}
      />
    )

    act(() => {
      wrapper.find('.button').simulate('click')
    })

    wrapper.update()

    expect(wrapper.find('.button').getDOMNode().className).toEqual('button active')
    expect(wrapper).toMatchSnapshot()
  })

  test('select / unselect', () => {
    const wrapper = mount(
      <Dropdown
        items={[{ id: '1', value: 'first' }, { id: '2', value: 'second' }]}
        title="Select something"
        onSelect={item => console.log('selected', item)}
      />
    )

    act(() => {
      wrapper.find('.button').simulate('click')
    })

    wrapper.update()

    expect(wrapper).toMatchSnapshot()

    const item = wrapper.find('.item').first()

    act(() => {
      item.simulate('click')
    })

    wrapper.update()

    expect(
      wrapper
        .find(Checkbox)
        .first()
        .prop('selected')
    ).toBeTruthy()
  })

  test('multi select', () => {
    const wrapper = mount(
      <Dropdown
        items={[{ id: '1', value: 'first' }, { id: '2', value: 'second' }]}
        title="Select something"
        onSelect={item => console.log('selected', item)}
        multi
      />
    )

    act(() => {
      wrapper.find('.button').simulate('click')
    })

    wrapper.update()

    expect(wrapper).toMatchSnapshot()

    const item = wrapper.find('.item').first()

    act(() => {
      item.simulate('click')
    })

    wrapper.update()

    expect(
      wrapper
        .find(Checkbox)
        .first()
        .prop('selected')
    ).toBeTruthy()

    act(() => {
      item.simulate('click')
    })

    wrapper.update()

    expect(
      wrapper
        .find(Checkbox)
        .first()
        .prop('selected')
    ).toBeFalsy()
  })
})
