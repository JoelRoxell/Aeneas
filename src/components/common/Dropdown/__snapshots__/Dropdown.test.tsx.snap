// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown multi select 1`] = `
<Dropdown
  items={
    Array [
      Object {
        "id": "1",
        "value": "first",
      },
      Object {
        "id": "2",
        "value": "second",
      },
    ]
  }
  multi={true}
  onSelect={[Function]}
  title="Select something"
>
  <div
    className="dropdown"
    onMouseLeave={[Function]}
  >
    <div
      className="button active"
      onClick={[Function]}
    >
      Select something
    </div>
    <div
      className="list active"
    >
      <div
        className="itemWrapper"
        key="0"
      >
        <div
          className="item"
          onClick={[Function]}
        >
          <Checkbox
            selected={false}
          >
            <div
              className="checkbox"
            >
              <input
                checked={false}
                className="checkbox"
                type="checkbox"
              />
            </div>
          </Checkbox>
          <div
            className="text"
          >
            first
          </div>
        </div>
      </div>
      <div
        className="itemWrapper"
        key="1"
      >
        <div
          className="item"
          onClick={[Function]}
        >
          <Checkbox
            selected={false}
          >
            <div
              className="checkbox"
            >
              <input
                checked={false}
                className="checkbox"
                type="checkbox"
              />
            </div>
          </Checkbox>
          <div
            className="text"
          >
            second
          </div>
        </div>
      </div>
    </div>
  </div>
</Dropdown>
`;

exports[`Dropdown render 1`] = `
<Dropdown
  items={
    Array [
      Object {
        "id": "1",
        "value": "first",
      },
      Object {
        "id": "2",
        "value": "second",
      },
    ]
  }
  onSelect={[Function]}
  title="Select something"
>
  <div
    className="dropdown"
    onMouseLeave={[Function]}
  >
    <div
      className="button"
      onClick={[Function]}
    >
      Select something
    </div>
    <div
      className="list"
    >
      <div
        className="itemWrapper"
        key="0"
      >
        <div
          className="item"
          onClick={[Function]}
        >
          <Checkbox
            selected={false}
          >
            <div
              className="checkbox"
            >
              <input
                checked={false}
                className="checkbox"
                type="checkbox"
              />
            </div>
          </Checkbox>
          <div
            className="text"
          >
            first
          </div>
        </div>
      </div>
      <div
        className="itemWrapper"
        key="1"
      >
        <div
          className="item"
          onClick={[Function]}
        >
          <Checkbox
            selected={false}
          >
            <div
              className="checkbox"
            >
              <input
                checked={false}
                className="checkbox"
                type="checkbox"
              />
            </div>
          </Checkbox>
          <div
            className="text"
          >
            second
          </div>
        </div>
      </div>
    </div>
  </div>
</Dropdown>
`;

exports[`Dropdown select / unselect 1`] = `
<Dropdown
  items={
    Array [
      Object {
        "id": "1",
        "value": "first",
      },
      Object {
        "id": "2",
        "value": "second",
      },
    ]
  }
  onSelect={[Function]}
  title="Select something"
>
  <div
    className="dropdown"
    onMouseLeave={[Function]}
  >
    <div
      className="button active"
      onClick={[Function]}
    >
      Select something
    </div>
    <div
      className="list active"
    >
      <div
        className="itemWrapper"
        key="0"
      >
        <div
          className="item"
          onClick={[Function]}
        >
          <Checkbox
            selected={false}
          >
            <div
              className="checkbox"
            >
              <input
                checked={false}
                className="checkbox"
                type="checkbox"
              />
            </div>
          </Checkbox>
          <div
            className="text"
          >
            first
          </div>
        </div>
      </div>
      <div
        className="itemWrapper"
        key="1"
      >
        <div
          className="item"
          onClick={[Function]}
        >
          <Checkbox
            selected={false}
          >
            <div
              className="checkbox"
            >
              <input
                checked={false}
                className="checkbox"
                type="checkbox"
              />
            </div>
          </Checkbox>
          <div
            className="text"
          >
            second
          </div>
        </div>
      </div>
    </div>
  </div>
</Dropdown>
`;

exports[`Dropdown show / hide 1`] = `
<Dropdown
  items={
    Array [
      Object {
        "id": "1",
        "value": "first",
      },
      Object {
        "id": "2",
        "value": "second",
      },
    ]
  }
  onSelect={[Function]}
  title="Select something"
>
  <div
    className="dropdown"
    onMouseLeave={[Function]}
  >
    <div
      className="button active"
      onClick={[Function]}
    >
      Select something
    </div>
    <div
      className="list active"
    >
      <div
        className="itemWrapper"
        key="0"
      >
        <div
          className="item"
          onClick={[Function]}
        >
          <Checkbox
            selected={false}
          >
            <div
              className="checkbox"
            >
              <input
                checked={false}
                className="checkbox"
                type="checkbox"
              />
            </div>
          </Checkbox>
          <div
            className="text"
          >
            first
          </div>
        </div>
      </div>
      <div
        className="itemWrapper"
        key="1"
      >
        <div
          className="item"
          onClick={[Function]}
        >
          <Checkbox
            selected={false}
          >
            <div
              className="checkbox"
            >
              <input
                checked={false}
                className="checkbox"
                type="checkbox"
              />
            </div>
          </Checkbox>
          <div
            className="text"
          >
            second
          </div>
        </div>
      </div>
    </div>
  </div>
</Dropdown>
`;
